src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace/RazorWorkspaceListener.cs

```diff
@@ -150,43 +134,37 @@ private void EnqueueUpdate(Project? project)
         }
 
         var projectId = project.Id;
-        if (_workQueues.TryGetValue(projectId, out var scheduler))
-        {
-            _logger.LogTrace("{projectId} scheduling task due to workspace event", projectId);
-
-            scheduler.ScheduleAsyncTask(ct => SerializeProjectAsync(projectId, ct), CancellationToken.None);
-        }
+        _workQueue.AddWork(projectId);
     }
 
-    // Protected for testing
-    protected virtual Task SerializeProjectAsync(ProjectId projectId, CancellationToken ct)
+    private async ValueTask UpdateCurrentProjectsAsync(ImmutableArray<ProjectId> projectIds, CancellationToken cancellationToken)
     {
-        if (_projectInfoFileName is null || _workspace is null)
-        {
-            return Task.CompletedTask;
-        }
+        var solution = _workspace.AssumeNotNull().CurrentSolution;
 
-        var project = _workspace.CurrentSolution.GetProject(projectId);
-        if (project is null)
+        foreach (var projectId in projectIds)
         {
-            return Task.CompletedTask;
-        }
+            if (_disposeTokenSource.IsCancellationRequested)
+            {
+                return;
+            }
 
-        _logger.LogTrace("{projectId} writing json file", projectId);
-        return RazorProjectInfoSerializer.SerializeAsync(project, _projectInfoFileName, ct);
+            await SerializeProjectAsync(projectId, solution, cancellationToken).ConfigureAwait(false);
+        }
     }
 
-    public void Dispose()
+    // Protected for testing
+#pragma warning disable RS0016 // Add public types and members to the declared API
```

Comment:
Suppressing until I can figure out if it's user error or analyzer error.

Reply:
Seems like neither. A protected method of a public type is part of the public API. I'd either remove the suppression, or create a TestAccessor so it doesn't need to be protected.

Reply:
Is this a case where `private protected` is needed?

Reply:
Yes! That's exactly what I was thinking. I just got my newfangled things mixed up